compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs

```diff
@@ -5108,6 +5108,44 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
             _ => {}
         }
     }
+    pub(crate) fn suggest_impl_similarly_named_trait(
+        &self,
+        err: &mut Diag<'_>,
+        obligation: &PredicateObligation<'tcx>,
+        trait_predicate: ty::PolyTraitPredicate<'tcx>,
+    ) {
+        let trait_def_id = trait_predicate.def_id();
+        let trait_name = self.tcx.item_name(trait_def_id);
+        if let Some(other_trait_def_id) = self.tcx.all_traits_including_private().find(|def_id| {
+            trait_def_id != *def_id
+                && trait_name == self.tcx.item_name(def_id)
+                && self
+                    .tcx
+                    .all_impls(*def_id)
+                    .find(|impl_def_id| {
+                        let Some(trait_ref) = self.tcx.impl_trait_ref(impl_def_id) else {
+                            return false;
+                        };
+                        self.predicate_must_hold_modulo_regions(&Obligation::new(
+                            self.tcx,
+                            obligation.cause.clone(),
+                            obligation.param_env,
+                            trait_ref
+                                .skip_binder()
+                                .with_self_ty(self.tcx, trait_predicate.skip_binder().self_ty()),
```

Comment:
`skip_binder` is pretty much never correct:

https://rustc-dev-guide.rust-lang.org/ty_module/early_binder.html?highlight=Binder#earlybinder-and-instantiating-parameters

also see https://rustc-dev-guide.rust-lang.org/typing_parameter_envs.html

U will also need to properly handle the generic parameters of the traits here, e.g. what if we've got `trait Trait {}` and a separate `trait Trait<T> {}`

Also, check predicate must hold for the other trait, not per impl.

Reply:
Mhhhh, I see. I have to document myself about this part, thanks for the links.

> Also, check predicate must hold for the other trait, not per impl.
Would you have a usecase for this ? You're probably right, but I don't get it.

thanks !